File: GFG/BST/check.cpp
#include<iostream>
#include"../ds.h"
using namespace std;

int main() {
    BST<int> bst;

    // Insertion
    bst.insert(50);
    bst.insert(30);
    bst.insert(70);
    bst.insert(20);
    bst.insert(40);
    bst.insert(60);
    bst.insert(80);
    bst.levelorder();
    // Search
    cout << "Searching for value 40: " << (bst.search(40) ? "Found" : "Not found") << endl;
    cout << "Searching for value 90: " << (bst.search(90) ? "Found" : "Not found") << endl;

    // Removal
    bst.remove(40);

    // Traversals
    cout << "Inorder traversal: ";
    bst.inorder();
    cout << endl;

    cout << "Preorder traversal: ";
    bst.preorder();
    cout << endl;

    cout << "Postorder traversal: ";
    bst.postorder();
    cout << endl;
    bst.levelorder();


    // Height
    cout << "Tree height: " << bst.height() << endl;

    // Size
    cout << "Tree size: " << bst.size() << endl;

    // Check if tree is empty
    cout << "Is tree empty? " << (bst.isEmpty() ? "Yes" : "No") << endl;

    // Floor and Ceil
    BNode<int>* floorNode = bst.floor(35);
    BNode<int>* ceilNode = bst.ceil(55);
    cout << "Floor of 35: " << (floorNode ? to_string(floorNode->key) : "Not found") << endl;
    cout << "Ceil of 55: " << (ceilNode ? to_string(ceilNode->key) : "Not found") << endl;

    return 0;
}

File: GFG/BST/check.py
import ds

bst=ds.BST()

bst.insert(50)
bst.insert(30)
bst.insert(70)
bst.insert(20)
bst.insert(40)
bst.insert(60)
bst.insert(80)

print("ROOT is : ",bst.root.key)

# traversals
bst.levelorder()
print("inorder traversal")
bst.inorder()
print("Preorder traversal")
bst.preorder()
print("Postorder traversal")
bst.postorder()
print()

# search

print("Searching for 40: ")
print("Found") if(bst.search(40)) else print("Not found")
print("Searching for 90: ")
print("Found") if(bst.search(90)) else print("Not found")

bst.delete(40)
print("40 deleted")
bst.levelorder()
# height and size
print("BST height = ",bst.height())
print("BST size = ",bst.size())

#emplty floor and ceil
print("Is BS empty? ",bst.isEmpty())
print("Floor of 35 = ",bst.floor(35).key)
print("Ceil of 55 = ",bst.ceil(55).key)


